nonnative/src/allocated_nonnative_field_var.rs

```diff
@@ -0,0 +1,982 @@
+                        .add_constant(cs.ns(|| format!("add_constant_{}", i)), &temp)?
+                        .sub(cs.ns(|| format!("sub_{}", i)), &other_limb)?,
+                );
+            } else {
+                let temp = pad_top_limb + &*pad_to_kp_limb;
+                limbs.push(
+                    this_limb
+                        .add_constant(cs.ns(|| format!("add_constant_{}", i)), &temp)?
+                        .sub(cs.ns(|| format!("sub_{}", i)), &other_limb)?,
+                );
+            }
+        }
+
+        let result = AllocatedNonNativeFieldVar::<TargetField, BaseField> {
+            limbs,
+            num_of_additions_over_normal_form: self.num_of_additions_over_normal_form
+                + &(other.num_of_additions_over_normal_form + &BaseField::one())
+                + &(other.num_of_additions_over_normal_form + &BaseField::one()),
+            is_in_the_normal_form: false,
+            target_phantom: PhantomData,
+        };
+
+        Ok(result)
+    }
+
+    /// Subtract a nonnative field element
+    pub fn sub<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS, other: &Self) -> Result<Self, SynthesisError> {
+        assert_eq!(self.get_optimization_type(), other.get_optimization_type());
+
+        let mut result = self.sub_without_reduce(&mut cs.ns(|| "sub_without_reduce"), other)?;
+        Reducer::<TargetField, BaseField>::post_add_reduce(cs, &mut result)?;
+        Ok(result)
+    }
+
+    /// Subtract a constant
+    pub fn sub_constant<CS: ConstraintSystem<BaseField>>(
+        &self,
+        cs: &mut CS,
+        other: &TargetField,
+    ) -> Result<Self, SynthesisError> {
+        let constant = Self::constant(&mut cs.ns(|| "constant"), *other)?;
+
+        self.sub(&mut cs.ns(|| "sub"), &constant)
+    }
+
+    /// Multiply a nonnative field element
+    pub fn mul<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS, other: &Self) -> Result<Self, SynthesisError> {
+        assert_eq!(self.get_optimization_type(), other.get_optimization_type());
+
+        self.mul_without_reduce(cs, &other)?.reduce(&mut cs.ns(|| "reduce"))
+    }
+
+    /// Multiply a constant
+    pub fn mul_constant<CS: ConstraintSystem<BaseField>>(
+        &self,
+        cs: &mut CS,
+        other: &TargetField,
+    ) -> Result<Self, SynthesisError> {
+        let constant = Self::constant(&mut cs.ns(|| "constant"), *other)?;
+
+        self.mul(&mut cs.ns(|| "mul"), &constant)
+    }
+
+    /// Compute the negate of a nonnative field element
+    pub fn negate<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS) -> Result<Self, SynthesisError> {
+        let zero = Self::zero(&mut cs.ns(|| "zero"))?;
+        zero.sub(&mut cs.ns(|| "sub"), self)
+    }
+
+    /// Compute the inverse of a nonnative field element
+    pub fn inverse<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS) -> Result<Self, SynthesisError> {
+        let inverse = Self::alloc_input(&mut cs.ns(|| "alloc_input"), || {
```

Comment:
Should this be `alloc` instead of `alloc_input`?

Reply:
Yep should be changed. The original notation was `new_witness` which should be equivalent to our `alloc`.

Reply:
https://github.com/AleoHQ/snarkVM/pull/78/commits/a4f346c81761a60457a8d5d935a696950f39b6d1